server/resp/src/main/java/org/infinispan/server/resp/commands/json/JSONSET.java

```diff
@@ -0,0 +1,148 @@
+         .jsonProvider(new com.jayway.jsonpath.spi.json.JacksonJsonNodeJsonProvider())
+         .mappingProvider(new com.jayway.jsonpath.spi.mapper.JacksonMappingProvider())
+         .build();
+
+   public static void jsonSetBiConsumer(CharSequence value, ResponseWriter writer) {
+      if (value == null) {
+         // Use null consumer
+         writer.nulls();
+         return;
+      }
+      if (value.length() == 2 && value.charAt(0) == 'O' && value.charAt(1) == 'K') {
+         // Use ok consumer
+         writer.ok();
+         return;
+      }
+      // All the rest are errors
+      writer.error(value);
+   }
+
+   public JSONSET() {
+      super("JSON.SET", -3, 1, 1, 1);
+   }
+
+   @Override
+   public CompletionStage<RespRequestHandler> perform(Resp3Handler handler, ChannelHandlerContext ctx,
+         List<byte[]> arguments) {
+      byte[] key = arguments.get(0);
+      byte[] path = arguments.get(1);
+      byte[] value = arguments.get(2);
+      if (arguments.size() > 4) {
+         handler.writer().syntaxError();
+         return handler.myStage();
+      }
+      boolean nx, xx; // must be effectively final
+      if (arguments.size() == 4) {
+         String arg = (new String(arguments.get(3)).toUpperCase());
+         switch (arg) {
+            case NX:
+               nx = true;
+               xx = false;
+               break;
+            case XX:
+               xx = true;
+               nx = false;
+               break;
+            default:
+               handler.writer().syntaxError();
+               return handler.myStage();
+         }
+      } else {
+         nx = false;
+         xx = false;
+      }
+      JsonNode newNode;
+      try {
+         newNode = JSONUtil.objectMapper.readTree(value);
+      } catch (IOException e) {
+         throw new CacheException(e);
+      }
+      FunctionalMap.ReadWriteMap<byte[], JsonDoc> cache = ReadWriteMapImpl
+            .create(FunctionalMapImpl.create(handler.typedCache(null)));
+      CompletionStage<CharSequence> cs = cache.eval(key, view -> {
+         var doc = (JsonDoc) view.find().orElse(null);
+         if (doc == null) {
+            if (xx) {
+               return null;
+            }
+            if (!isRoot(path)) {
+               throw new CacheException("new objects must be created at root");
+            }
+            view.set(new JsonDoc(value));
+            return RespConstants.OK;
+         }
+         if (nx) {
+            return null;
+         }
+         if (isRoot(path)) {
+            // Updating the root node is not allowed by jsonpath
+            // replacing the whole doc here
+            view.set(new JsonDoc(value));
+            return RespConstants.OK;
+         }
+         try {
+            var rootObjectNode = (ObjectNode) JSONUtil.objectMapper.readTree(doc.bytesDocument());
+            var jpCtx = JsonPath.using(config).parse(rootObjectNode);
+            var pathStr = new String(path);
```

Comment:
Should we specify UTF-8?

Reply:
mmm yes I guess so, jsonpath default is utf-8. thanks!